File: sdks/dotnet/Condux.Sdk/EventTransport.cs
using System.Globalization;
using System.Net;
using System.Text;

namespace Condux.Sdk;

// POSTs a serialized event to the relay, retrying transient failures with backoff. Never throws (except
// on caller cancellation): it returns a SendResult describing the outcome.
//
// The retry schedule is identical to every other Condux SDK's, pinned by sdks/conformance/backoff.tsv,
// which BackoffConformanceTests drives through the public send path. The citation is the point: the
// claim is otherwise made in a comment and checked nowhere.
internal static class EventTransport
{
    private static readonly TimeSpan BaseBackoff = TimeSpan.FromMilliseconds(200);
    private static readonly TimeSpan MaxBackoff = TimeSpan.FromSeconds(30);

    public static async Task<SendResult> SendAsync(
        HttpClient http, string url, string publicKey, string body, int maxRetries,
        Func<TimeSpan, CancellationToken, Task> sleep, CancellationToken cancellationToken)
    {
        int? lastStatus = null;
        string? lastError = null;

        for (var attempt = 0; attempt <= maxRetries; attempt++)
        {
            HttpResponseMessage? response = null;
            try
            {
                using var request = new HttpRequestMessage(HttpMethod.Post, url)
                {
                    Content = new StringContent(body, Encoding.UTF8, "application/json"),
                };
                request.Headers.TryAddWithoutValidation("x-condux-auth", publicKey);
                response = await http.SendAsync(request, cancellationToken);
            }
            catch (HttpRequestException ex)
            {
                lastError = ex.Message;
            }
            catch (TaskCanceledException ex) when (!cancellationToken.IsCancellationRequested)
            {
                lastError = ex.Message; // a request timeout, not caller cancellation
            }

            if (response is not null)
            {
                lastStatus = (int)response.StatusCode;
                lastError = null;
                if (response.IsSuccessStatusCode)
                {
                    response.Dispose();
                    return new SendResult(true, attempt + 1, lastStatus);
                }

                if (!IsRetriable(response.StatusCode))
                {
                    response.Dispose();
                    return new SendResult(false, attempt + 1, lastStatus);
                }
            }

            if (attempt == maxRetries)
            {
                response?.Dispose();
                break;
            }

            await sleep(Backoff(attempt, response), cancellationToken);
            response?.Dispose();
        }

        return new SendResult(false, maxRetries + 1, lastStatus, lastError);
    }

    // 429 (rate limited) and 5xx are worth retrying; other 4xx (bad DSN, bad payload) are not.
    private static bool IsRetriable(HttpStatusCode status) =>
        status == HttpStatusCode.TooManyRequests || (int)status >= 500;

    // Honor Retry-After (seconds) on a 429, else exponential backoff. Both paths are capped at
    // MaxBackoff, at ONE return so a later branch cannot route past it: the SDK holds the caller's
    // thread while it waits, so bounding that wait is its own obligation and not the relay's to set.
    private static TimeSpan Backoff(int attempt, HttpResponseMessage? response)
    {
        var requested = response?.StatusCode == HttpStatusCode.TooManyRequests
            ? ParseRetryAfter(response)
            : null;
        var milliseconds = requested?.TotalMilliseconds
                           ?? BaseBackoff.TotalMilliseconds * Math.Pow(2, attempt);
        return TimeSpan.FromMilliseconds(Math.Min(milliseconds, MaxBackoff.TotalMilliseconds));
    }

    // Read Retry-After as the string it arrived as, not through Headers.RetryAfter. That typed property
    // exposes a Delta only for RFC 7231 delta-seconds, a non-negative INTEGER, so it cannot represent a
    // fractional value at all, and a proxy or a CDN in front of the relay is where fractional values come
    // from. Negative is clamped rather than discarded, so a sender merely wrong about the sign is read as
    // asking to retry now.
    private static TimeSpan? ParseRetryAfter(HttpResponseMessage response)
    {
        if (!response.Headers.TryGetValues("Retry-After", out var values))
        {
            return null;
        }

        // IsFinite is load-bearing, not defensive. TryParse accepts "Infinity" and "NaN", and
        // TimeSpan.FromSeconds throws OverflowException on the first and ArgumentException on the
        // second, out of a transport whose whole contract is that it never throws. Not finite is not
        // an instruction.
        var raw = values.FirstOrDefault();
        return double.TryParse(raw, NumberStyles.Float, CultureInfo.InvariantCulture, out var seconds)
               && double.IsFinite(seconds)
            ? TimeSpan.FromSeconds(Math.Max(0, seconds))
            : null;
    }
}

File: sdks/ruby/lib/condux/event_transport.rb
# frozen_string_literal: true

require "net/http"
require "uri"
require_relative "send_result"

module Condux
  # Delivers a serialized event to the relay, retrying transient failures (429 / 5xx / network) with
  # capped exponential backoff, honoring Retry-After on a 429. Never raises — returns a SendResult.
  class EventTransport
    BASE_BACKOFF_MS = 200.0
    MAX_BACKOFF_MS = 30_000.0

    def initialize(store_url, public_key, max_retries, transport, sleep)
      @store_url = store_url
      @public_key = public_key
      @max_retries = max_retries
      @transport = transport || method(:net_http_send)
      @sleep = sleep || ->(seconds) { Kernel.sleep(seconds) }
    end

    def send_event(body)
      last_status = nil
      last_error = nil
      attempt = 0
      while attempt <= @max_retries
        status, headers, error = attempt_send(body)
        if error
          last_error = error
        else
          last_status = status
          last_error = nil
          return SendResult.new(ok: true, attempts: attempt + 1, status: status) if (200..299).cover?(status)
          return SendResult.new(ok: false, attempts: attempt + 1, status: status) unless retriable?(status)
        end

        break if attempt == @max_retries

        @sleep.call(backoff_seconds(attempt, status, headers))
        attempt += 1
      end

      SendResult.new(ok: false, attempts: @max_retries + 1, status: last_status, error: last_error)
    end

    private

    def attempt_send(body)
      headers = { "Content-Type" => "application/json", "x-condux-auth" => @public_key }
      status, response_headers = @transport.call(@store_url, headers, body)
      [status, response_headers || {}, nil]
    rescue StandardError => e # a genuine network failure — retry; a failed send must never crash the caller
      message = e.message.to_s
      [nil, {}, message.empty? ? e.class.name : message]
    end

    # 429 (rate limited) and 5xx are worth retrying; other 4xx (bad DSN / payload) are not.
    def retriable?(status)
      status == 429 || status >= 500
    end

    # Honor Retry-After (seconds) on a 429, else exponential backoff. Both paths are capped at
    # MAX_BACKOFF_MS, at ONE return so a later branch cannot route past it: the SDK holds the caller's
    # thread while it waits, so bounding that wait is its own obligation and not the relay's to set.
    # Returns seconds.
    def backoff_seconds(attempt, status, headers)
      wanted_ms = BASE_BACKOFF_MS * (2**attempt)
      if status == 429
        retry_after = header(headers, "retry-after")
        if retry_after && !retry_after.strip.empty?
          seconds = begin
            Float(retry_after)
          rescue ArgumentError, TypeError
            nil
          end
          wanted_ms = [seconds, 0.0].max * 1000.0 if seconds
        end
      end
      [wanted_ms, MAX_BACKOFF_MS].min / 1000.0
    end

    def header(headers, name)
      target = name.downcase
      headers.each { |key, value| return value if key.to_s.downcase == target }
      nil
    end

    def net_http_send(url, headers, body)
      uri = URI(url)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.scheme == "https"
      request = Net::HTTP::Post.new(uri)
      headers.each { |key, value| request[key] = value }
      request.body = body
      response = http.request(request)
      [response.code.to_i, response]
    end
  end
end
